docs/configuration.md
---
title: Configuration
---
# Configuration

System-wide configuration parameters can be set by creating a ConfigMap with name `model-serving-config`. It should contain a single key named `config.yaml`, whose value is a yaml doc containing the configuration. All parameters have defaults and are optional. If the ConfigMap does not exist, all parameters will take their defaults.

The configuration can be updated at runtime and will take effect immediately. Be aware however that certain changes could cause temporary disruption to the service - in particular changing the service name, port, TLS configuration and/or headlessness.

Example:

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: model-serving-config
data:
  config.yaml: |
    #Sample config overrides
    inferenceServiceName: "modelmesh-serving"
    inferenceServicePort: 8033
    podsPerRuntime: 2
    metrics:
      enabled: true
```

The following parameters are currently supported. _Note_ the keys are expressed here in camel case but are in fact case-insensitive.

| Variable                         | Description                                                                                             | Default              |
| -------------------------------- | ------------------------------------------------------------------------------------------------------- | -------------------- |
| `inferenceServiceName`           | The service name which is used for communication with the serving server                                | `modelmesh-serving`        |
| `inferenceServicePort`           | The port number for communication with the inferencing service                                          | `8033`               |
| `storageSecretName`              | The secret containing entries for each storage backend from which models can be loaded (\* see below) | `storage-config`     |
| `podsPerRuntime`                 | Number of server Pods to run per enabled Serving Runtime (\*\* see below)                             | `2`                  |
| `tls.secretName`                 | Kubernetes TLS type secret to use for securing the Service; no TLS if empty (\*\*\* see below)        |                      |
| `tls.clientAuth`                 | Enables mutual TLS authentication. Supported values are `required` and `optional`, disabled if empty    |                      |
| `headlessService`                | Whether the Service should be headless (recommended)                                                    | `true`               |
| `enableAccessLogging`            | Enables logging of each request to the model server                                                     | `false`              |
| `serviceAccountName`             | The service account to use for runtime Pods                                                             | `modelmesh`         |
| `metrics.enabled`                | Enables serving of Prometheus metrics                                                                   | `true`               |
| `metrics.port`                   | Port on which to serve metrics via the `/metrics` endpoint                                              | `2112`               |
| `metrics.scheme`                 | Scheme to use for the `/metrics` endpoint (`http` or `https`)                                           | `https`              |
| `scaleToZero.enabled`            | Whether to scale down Serving Runtimes that have no Predictors                                          | `true`               |
| `scaleToZero.gracePeriodSeconds` | The number of seconds to wait after Predictors are deleted before scaling to zero                       | `60`                 |



(\*) Currently requires a controller restart to take effect

(\*\*) This parameter will likely be removed in a future release; the Pod replica counts will become more dynamic.

(\*\*\*) The TLS configuration secret allows for keys:

- `tls.crt` - path to TLS secret certificate
- `tls.key` - path to TLS secret key
- `ca.crt` (optional) - single path or comma-separated list of paths to trusted certificates

## Generating TLS Certificates for Dev/Test

TLS is enabled through adding a value for `tls.secretName` in the user's ConfigMap that points to an existing kube secret with TLS key/cert details.

To create a SAN key/cert for TLS, use command:

```sh
$ openssl req -x509 -newkey rsa:4096 -sha256 -days 3560 -nodes -keyout example.key -out example.crt -subj '/CN=modelmesh-serving' -extensions san -config openssl-san.config
```

Where the contents of `openssl-san.config` look like:

```
[ req ]
distinguished_name = req
[ san ]
subjectAltName = DNS:modelmesh-serving,DNS:localhost,IP:0.0.0.0
```

With the generated key/cert, create a kube secret with contents like:

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: <NAME_OF_SECRET>
type: kubernetes.io/tls
stringData:
  tls.crt: <contents-of-example.crt>
  tls.key: <contents-of-example.key>
  ca.crt: <contents-of-example.crt>
```

For basic TLS, only the fields `tls.crt` and `tls.key` are needed in the kube secret. For mutual TLS, add `ca.crt` in the kube secret and set the configuration `tls.clientAuth` to `require` in the ConfigMap `model-serving-config`.



docs-bak/configuration.md
---
title: Configuration
---

System-wide configuration parameters can be set by creating a ConfigMap with name `model-serving-config`. It should contain a single key named `config.yaml`, whose value is a yaml doc containing the configuration. All parameters have defaults and are optional. If the ConfigMap does not exist, all parameters will take their defaults.

The configuration can be updated at runtime and will take effect immediately. Be aware however that certain changes could cause temporary disruption to the service - in particular changing the service name, port, TLS configuration and/or headlessness.

Example:

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: model-serving-config
data:
  config.yaml: |
    #Sample config overrides
    inferenceServiceName: "wml-serving"
    inferenceServicePort: 8033
    podsPerRuntime: 2
    metrics:
      enabled: true
```

The following parameters are currently supported. _Note_ the keys are expressed here in camel case but are in fact case-insensitive.

| Variable                         | Description                                                                                             | Default              |
| -------------------------------- | ------------------------------------------------------------------------------------------------------- | -------------------- |
| `etcdSecretName`                 | DEPRECATED - The secret which contains etcd connection information (\* see below)                       | `model-serving-etcd` |
| `inferenceServiceName`           | The service name which is used for communication with the serving server                                | `wml-serving`        |
| `inferenceServicePort`           | The port number for communication with the inferencing service                                          | `8033`               |
| `storageSecretName`              | The secret containing entries for each storage backend from which models can be loaded (\*\* see below) | `storage-config`     |
| `podsPerRuntime`                 | Number of server Pods to run per enabled Serving Runtime (\*\*\* see below)                             | `2`                  |
| `tls.secretName`                 | Kubernetes TLS type secret to use for securing the Service; no TLS if empty (\*\*\*\* see below)        |                      |
| `tls.clientAuth`                 | Enables mutual TLS authentication. Supported values are `required` and `optional`, disabled if empty    |                      |
| `headlessService`                | Whether the Service should be headless (recommended)                                                    | `true`               |
| `enableAccessLogging`            | Enables logging of each request to the model server                                                     | `false`              |
| `serviceAccountName`             | The service account to use for runtime Pods                                                             | `wmlserving`         |
| `metrics.enabled`                | Enables serving of Prometheus metrics                                                                   | `true`               |
| `metrics.port`                   | Port on which to serve metrics via the `/metrics` endpoint                                              | `2112`               |
| `metrics.scheme`                 | Scheme to use for the `/metrics` endpoint (`http` or `https`)                                           | `https`              |
| `scaleToZero.enabled`            | Whether to scale down Serving Runtimes that have no Predictors                                          | `true`               |
| `scaleToZero.gracePeriodSeconds` | The number of seconds to wait after Predictors are deleted before scaling to zero                       | `60`                 |

(\*) The `etcdSecretName` variable has been deprecated and moved to an environment variable set by the operator. The proper way to modify this value is to set `spec.etcd.externalConnection.secretName` in the `WmlServing` resource.
The etcd configuration secret must contain a key named `etcd_connection` and have contents that conforms to https://github.com/IBM/etcd-java/blob/master/etcd-json-schema.md. If the etcd server is shared between namespaces and/or environments, ensure that the `root_prefix` json attribute is set to something unique per namespace.

(\*\*) Currently requires a controller restart to take effect

(\*\*\*) This parameter will likely be removed in a future release; the Pod replica counts will become more dynamic.

(\*\*\*\*) The TLS configuration secret allows for keys:

- `tls.crt` - path to TLS secret certificate
- `tls.key` - path to TLS secret key
- `ca.crt` (optional) - single path or comma-separated list of paths to trusted certificates

## Defining roles based on personas

1. The `Cluster-Admin`

   Installs wml-serving operator. This is top level role with widest privileges among all the personas.

2. The `Namespace-Admin` or `WMLServing-Admin`

   This persona can install and configure Model-Mesh Serving, `ServingRuntime`s, `ConfigMap`s and storage `Secrets`.

3. The `Model-Deployer`

   User pertaining to this persona can create, update and delete `Predictor`s.

There could be more personas or change in scope of existing personas as wml-serving operator evolves in future.

## Generating TLS Certificates for Dev/Test

TLS is enabled through adding a value for `tls.secretName` in the user's ConfigMap that points to an existing kube secret with TLS key/cert details.

To create a SAN key/cert for TLS, use command:

```sh
$ openssl req -x509 -newkey rsa:4096 -sha256 -days 3560 -nodes -keyout example.key -out example.crt -subj '/CN=wml-serving' -extensions san -config openssl-san.config
```

Where the contents of `openssl-san.config` look like:

```
[ req ]
distinguished_name = req
[ san ]
subjectAltName = DNS:wml-serving,DNS:localhost,IP:0.0.0.0
```

With the generated key/cert, create a kube secret with contents like:

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: <NAME_OF_SECRET>
type: kubernetes.io/tls
stringData:
  tls.crt: <contents-of-example.crt>
  tls.key: <contents-of-example.key>
  ca.crt: <contents-of-example.crt>
```

For basic TLS, only the fields `tls.crt` and `tls.key` are needed in the kube secret. For mutual TLS, add `ca.crt` in the kube secret and set the configuration `tls.clientAuth` to `require` in the ConfigMap `model-serving-config`.

## CloudPak Billing Annotations

CloudPak deployments require pod annotations to calculate proper billing. When Model-Mesh Serving is adopted inside a product, the product needs to provide these annotations.

- conversionRatio
- cloudpakId
- cloudpakName

These annotations will be provided inside a `ConfigMap` that must be deployed in the same namespace as the Model-Mesh Serving controller and runtimes. For example:

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: wml-product
data:
  annotations: |
    conversionRatio=2:3
    cloudpakId=8b5161b0-a57e-490b-835d-aa2e6b70b28b
    cloudpakName=IBM Cloud Pak for Data
```

Patch the existing serving runtime with the name of the `ConfigMap` in the `metadata.annotations`.

`kubectl patch servingruntime msp-ml-server-0.x -p '[{"op": "add", "path": "/metadata/annotations/productConfig", "value": "wml-product"}]' --type json`

Notice that `productConfig` annotation value must match with the actual `ConfigMap` name.

<InlineNotification>

NOTE: `conversionRatio` will default to `1:1`.

</InlineNotification>

For more details see the [billing annotations ADR](https://github.ibm.com/ai-foundation/ai-foundation/blob/main/docs/adr/008-cloudpak-billing-annotations.md).

### Red Hat OpenShift Security Context Constraints Requirements

Security context constraints (SCC) allows the administrators to control permissions for pods.

The operator uses a 'restricted' SCC. A custom SCC can be used with the operator and is the most restrictive SCC the operator can run with. Using this SCC, the operator can be independant of the clusters default restricted SCC.

- You could use below YAML snippet to enable the custom SecurityContextConstraints

Custom SecurityContextConstraints definition:

```yaml
apiVersion: security.openshift.io/v1
kind: SecurityContextConstraints

metadata:
  name: wmlserving-operator-custom-scc
  annotations:
    kubernetes.io/description:
      "This policy is the most restrictive for wmlserving operator,
      requiring pods to run with a non-root UID, and preventing pods from accessing the host.
      The UID and GID will be bound by ranges specified at the Namespace level."

allowHostDirVolumePlugin: false
allowHostIPC: false
allowHostNetwork: false
allowHostPID: false
allowHostPorts: false
allowPrivilegeEscalation: true
allowPrivilegedContainer: false
allowedCapabilities: null

defaultAddCapabilities: null
fsGroup:
  type: MustRunAs
users: []
groups: []
priority: 0
readOnlyRootFilesystem: false
requiredDropCapabilities:
  - ALL
runAsUser:
  type: MustRunAsRange
seLinuxContext:
  type: MustRunAs
supplementalGroups:
  type: RunAsAny
volumes:
  - configMap
  - downwardAPI
  - emptyDir
  - persistentVolumeClaim
  - projected
  - secret
```

- Restricted scc role definition:

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  name: wmlserving-controller-restricted-scc
rules:
  - apiGroups:
      - security.openshift.io
    resources:
      - securitycontextconstraints
    resourceNames:
      - wmlserving-operator-custom-scc
    verbs:
      - use
```

For each ServiceAccount (SA) that the SecurityContextConstraint (SCC) needs to be associated with, copy and paste the following snippet into the 'Import YAML' page of the OpenShift user interface, replacing the subjects.name with the relevant ServiceAccount name. For example, the wmlserving-operator ServiceAccount name.

- Restricted scc rolebinding definition:

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: controller-restricted-scc
subjects:
  - kind: ServiceAccount
    name: wmlserving-operator
roleRef:
  kind: Role
  name: wmlserving-controller-restricted-scc
  apiGroup: rbac.authorization.k8s.io
```


controllers/config.go
// Copyright 2021 IBM Corporation
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
package controllers

import (
	"context"
	"fmt"
	"os"
	"path"
	"runtime"
	"strings"
	"sync"
	"sync/atomic"
	"unsafe"

	"sigs.k8s.io/controller-runtime/pkg/handler"
	"sigs.k8s.io/controller-runtime/pkg/reconcile"

	"github.com/spf13/viper"
	corev1 "k8s.io/api/core/v1"
	"k8s.io/apimachinery/pkg/api/errors"
	"k8s.io/apimachinery/pkg/api/resource"
	"k8s.io/apimachinery/pkg/types"
	ctrl "sigs.k8s.io/controller-runtime"
	"sigs.k8s.io/controller-runtime/pkg/client"
)

const (
	EnvEtcdSecretName     = "ETCD_SECRET_NAME"
	DefaultEtcdSecretName = "model-serving-etcd"
)

var (
	defaultConfig *viper.Viper
	configLog     = ctrl.Log.WithName("config")
)

// Type Config holds process global configuration information
type Config struct {
	// System config
	EtcdSecretName    string // DEPRECATED - should be removed in the future
	ModelMeshEndpoint string // For dev use only

	// Service config
	InferenceServiceName string
	InferenceServicePort uint16
	TLS                  TLSConfig
	HeadlessService      bool

	// Runtimes config
	ModelMeshImage         ImageConfig
	ModelMeshResources     ResourceRequirements
	StorageHelperImage     ImageConfig
	StorageHelperResources ResourceRequirements
	PodsPerRuntime         uint16
	StorageSecretName      string
	EnableAccessLogging    bool

	ServiceAccountName string

	Metrics     PrometheusConfig
	ScaleToZero ScaleToZeroConfig
}

type PrometheusConfig struct {
	Enabled bool
	Port    uint16
	Scheme  string
}

type ScaleToZeroConfig struct {
	Enabled bool
	// how long to wait after the last predictor assigned to a runtime is deleted
	// before scaling to zero
	GracePeriodSeconds uint16
}

type TLSConfig struct {
	// TLS disabled if omitted
	SecretName string
	// Mutual TLS disabled if omitted
	ClientAuth string
}

func (c *Config) GetEtcdSecretName() string {
	secretName, found := os.LookupEnv(EnvEtcdSecretName)
	if !found {
		secretName = DefaultEtcdSecretName
	}

	// for backward compatability with the old configmap - should be removed in the future
	if secretName == DefaultEtcdSecretName && c.EtcdSecretName != "" {
		secretName = c.EtcdSecretName
	}

	return secretName
}

// Type Config holds process global configuration information
type ConfigProvider struct {
	config                unsafe.Pointer
	c                     sync.Cond
	isReloading           bool
	loadedResourceVersion string
}

func NewConfigProvider(ctx context.Context, cl client.Client, name types.NamespacedName) (*ConfigProvider, error) {
	// Perform initial load of the default configuration
	config, err := NewMergedConfigFromString("")
	if err != nil {
		return nil, err
	}

	cp := &ConfigProvider{c: sync.Cond{L: &sync.Mutex{}}, config: (unsafe.Pointer)(config)}
	return cp, cp.ReloadConfigMap(ctx, cl, name)
}

func (cp *ConfigProvider) GetConfig() *Config {
	return (*Config)(atomic.LoadPointer(&cp.config))
}

func (cp *ConfigProvider) ReloadConfigMap(ctx context.Context, c client.Client, name types.NamespacedName) error {
	cp.c.L.Lock()
	cp.isReloading = true
	defer func() {
		cp.c.L.Unlock()
		cp.isReloading = false
	}()

	// get the user ConfigMap
	configmap := corev1.ConfigMap{}
	err := c.Get(ctx, name, &configmap)
	// any error other than NotFound is unexpected
	if err != nil && !errors.IsNotFound(err) {
		return err
	}

	// only reload if the resource version changed from what we last loaded successfully
	//  if the user config does not exist, configmap.ResourceVersion is the empty string
	if configmap.ResourceVersion != cp.loadedResourceVersion {
		var err2 error
		var newConfig *Config
		//  if the configmap was deleted but had been previously loaded
		if configmap.ResourceVersion == "" {
			configLog.Info("User configmap deleted, reverting to defaults", "ConfigMap", name)
			// load from empty string to by-pass validation of the configmap
			newConfig, err2 = NewMergedConfigFromString("")
		} else {
			configLog.Info("Reloading user config", "ConfigMap", name)
			newConfig, err2 = NewMergedConfigFromConfigMap(configmap)
		}
		if err2 != nil {
			return err2
		}
		// update the stored resource version to track changes
		atomic.StorePointer(&cp.config, (unsafe.Pointer)(newConfig))
		cp.loadedResourceVersion = configmap.ResourceVersion
	}

	cp.c.Broadcast()

	return nil
}

// Handler used by controllers which depend on the user configuration
func ConfigWatchHandler(configMapName types.NamespacedName, f func() []reconcile.Request,
	cp *ConfigProvider, kclient *client.Client) handler.EventHandler {
	return handler.EnqueueRequestsFromMapFunc(func(o client.Object) []reconcile.Request {
		// Ignore ConfigMaps we don't care about
		if o.GetName() == configMapName.Name && o.GetNamespace() == configMapName.Namespace {
			err := cp.ReloadConfigMap(context.TODO(), *kclient, configMapName)
			if err != nil {
				configLog.Error(err, "Unable to reload user configuration")
			}
			return f()
		}
		return []reconcile.Request{}
	})
}

func (cp *ConfigProvider) IsReloading() bool {
	return cp.isReloading
}

func (cp *ConfigProvider) AwaitReload() {
	cp.c.L.Lock()
	defer cp.c.L.Unlock()

	if cp.isReloading {
		cp.c.Wait()
	}
}

type ImageConfig struct {
	Name    string
	Tag     string
	Digest  string
	Command []string
}

func (m ImageConfig) TaggedImage() string {
	// Use only the digest if it is provided and ignore the tag.
	// It is possible to reference an image by both tag and digest, but this is considered ambiguous and the digest would be used to pull the image anyway.
	// See also https://github.com/cri-o/cri-o/pull/3060
	if m.Digest != "" {
		return fmt.Sprintf("%s@%s", m.Name, m.Digest)
	} else if m.Tag != "" {
		return fmt.Sprintf("%s:%s", m.Name, m.Tag)
	}
	return m.Name
}

type ResourceRequirements struct {
	Requests ResourceQuantities
	Limits   ResourceQuantities
	// used to cache the parsed resources from parseAndValidate()
	parsedKubeResourceRequirements *corev1.ResourceRequirements
}
type ResourceQuantities struct {
	CPU    string
	Memory string
}

func (rr ResourceRequirements) ToKubernetesType() *corev1.ResourceRequirements {
	// assert that parseAndValidate has already been called
	if rr.parsedKubeResourceRequirements == nil {
		panic("ResourceRequirements: Must call parseAndValidate() before ToKubernetesType()")
	}

	return rr.parsedKubeResourceRequirements
}

func (rr *ResourceRequirements) parseAndValidate() error {
	var err error
	var limitsCPU, requestsCPU, limitsMemory, requestsMemory resource.Quantity

	if limitsCPU, err = resource.ParseQuantity(rr.Limits.CPU); err != nil {
		return fmt.Errorf("Cannot parse 'Limits.CPU' as a quantity: %w", err)
	}
	if requestsCPU, err = resource.ParseQuantity(rr.Requests.CPU); err != nil {
		return fmt.Errorf("Cannot parse 'Requests.CPU' a quantity: %w", err)
	}

	if limitsMemory, err = resource.ParseQuantity(rr.Limits.Memory); err != nil {
		return fmt.Errorf("Cannot parse 'Limits.Memory' as a quantity: %w", err)
	}
	if requestsMemory, err = resource.ParseQuantity(rr.Requests.Memory); err != nil {
		return fmt.Errorf("Cannot parse 'Requests.Memory' a quantity: %w", err)
	}

	rr.parsedKubeResourceRequirements = &corev1.ResourceRequirements{
		Limits: corev1.ResourceList{
			corev1.ResourceCPU:    limitsCPU,
			corev1.ResourceMemory: limitsMemory,
		},
		Requests: corev1.ResourceList{
			corev1.ResourceCPU:    requestsCPU,
			corev1.ResourceMemory: requestsMemory,
		},
	}

	return nil
}

// Sets the defaults prior to config file parsing
// Note: see also config/default/defaults.yaml for the shipped default config
func defaults(v *viper.Viper) {
	v.SetDefault("InferenceServiceName", "modelmesh-serving")
	v.SetDefault("InferenceServicePort", 8033)
	v.SetDefault("PodsPerRuntime", 2)
	v.SetDefault("StorageSecretName", "storage-config")
	v.SetDefault("ServiceAccountName", "wmlserving")
	v.SetDefault("Metrics.Port", 2112)
	v.SetDefault("Metrics.Scheme", "https")
	v.SetDefault("ScaleToZero.Enabled", true)
	v.SetDefault("ScaleToZero.GracePeriodSeconds", 60)
}

func init() {
	defaultConfig = viper.New()

	defaults(defaultConfig)

	defaultConfig.SetConfigName("config-defaults")
	defaultConfig.SetConfigType(ConfigType)
	defaultConfig.AddConfigPath(MountLocation)

	//For dev env and tests, must get the config filename using cwd relative path
	if _, filename, _, ok := runtime.Caller(0); !ok {
		panic("Unable to get the caller")
	} else {
		filepath := path.Join(path.Dir(filename), "../config/default")
		defaultConfig.AddConfigPath(filepath)
	}

	if err := defaultConfig.ReadInConfig(); err != nil {
		configLog.Error(err, "Unable to read the default configuration", "path", MountLocation)
	}
}

func NewMergedConfigFromConfigMap(m corev1.ConfigMap) (*Config, error) {
	configYaml, ok := m.Data["config.yaml"]
	if !ok {
		err := fmt.Errorf("User ConfigMap must contain a key named config.yaml")
		return nil, err
	}

	return NewMergedConfigFromString(configYaml)
}

func NewMergedConfigFromString(configYaml string) (*Config, error) {
	var err error

	v := viper.New()
	v.SetConfigType(ConfigType)
	for _, key := range defaultConfig.AllKeys() {
		v.SetDefault(key, defaultConfig.Get(key))
	}

	configYamlReader := strings.NewReader(configYaml)
	if err = v.ReadConfig(configYamlReader); err != nil {
		return nil, err
	}

	// Even if the default config has an image digest, a user should be able to
	// override it with a tag (ignoring the default digest)
	// HACK: There should be a better way to do this...
	if v.GetString("storageHelperImage.tag") != defaultConfig.GetString("storageHelperImage.tag") &&
		v.GetString("storageHelperImage.digest") == defaultConfig.GetString("storageHelperImage.digest") {
		v.Set("storageHelperImage.digest", "")
	}
	if v.GetString("modelMeshImage.tag") != defaultConfig.GetString("modelMeshImage.tag") &&
		v.GetString("modelMeshImage.digest") == defaultConfig.GetString("modelMeshImage.digest") {
		v.Set("modelMeshImage.digest", "")
	}

	// unmarshal the config into a Config struct
	var config Config
	if err = v.Unmarshal(&config); err != nil {
		return nil, err
	}

	configLog.Info("Updated model serving config", "mergedConfig", config)

	// extra validations on parsed config
	if err = config.ModelMeshResources.parseAndValidate(); err != nil {
		return nil, fmt.Errorf("Invalid config for 'ModelMeshResources': %s", err)
	}
	if err = config.StorageHelperResources.parseAndValidate(); err != nil {
		return nil, fmt.Errorf("Invalid config for 'StorageHelperResources': %s", err)
	}

	return &config, nil
}
